modules/assignment/src/main/java/com/intuit/wasabi/assignment/impl/AssignmentsImpl.java

```diff
@@ -619,12 +616,26 @@ protected Assignment getAssignment(User.ID userID, Application.Name applicationN
                     Assignment.Status.EXPERIMENT_NOT_STARTED);
         }
 
+        Assignment assignment = getAssignment(experimentID, userID, context, userAssignments, bucketList);
+
+        // Case: Experiment expires -> existing Assignments are returned and new Assignments
+        // are returned as null assignments
         if (currentTime > experiment.getEndTime().getTime()) {
```

Comment:
The expire check now only updates the state of the experiment, but not only in the db but also on the object for the later changes. I also added an else condition for the case where the database holds a different state than the current experiment - not sure though what the best behavior is .. currently it sets the database state as the new state, but we could also throw an error.